src/value.rs

```diff
@@ -140,27 +330,76 @@ impl Value {
         self.ty.is_unit()
     }
 
+    /// Helper function to read the first bit of a value
+    ///
+    /// If the first bit is not available (e.g. if the value has zero size)
+    /// then returns None.
+    fn first_bit(&self) -> Option<bool> {
+        let mask = if self.bit_offset % 8 == 0 {
+            0x80
+        } else {
+            1 << (7 - self.bit_offset % 8)
+        };
+        let res = self
+            .inner
+            .get(self.bit_offset / 8)
+            .map(|x| x & mask == mask);
+        res
+    }
+
     /// Access the inner value of a left sum value.
     pub fn to_left(&self) -> Option<Self> {
-        match &self.inner {
-            ValueInner::Left(inner) => Some(inner.shallow_clone()),
-            _ => None,
+        if self.first_bit() == Some(false) {
+            if let Some((lty, rty)) = self.ty.as_sum() {
+                let sum_width = 1 + cmp::max(lty.bit_width(), rty.bit_width());
+                Some(Self {
+                    inner: Arc::clone(&self.inner),
+                    bit_offset: self.bit_offset + sum_width - lty.bit_width(),
```

Comment:
cbbe9b5a3e7982c5b4c85b28cc37996fbcb0c7b2: `bit_offset: self.bit_offset + self.ty.bit_width() - lty.bit_width()`

The variable `sum_width` is not required.

Reply:
ooo nice catch